_posts/2019-06-05-uncertainty-estimates-tfprobability/uncertainty_tfprobability.Rmd

```diff
@@ -0,0 +1,779 @@
+
+## Aleatoric vs. epistemic uncertainty
+
+Reminiscent somehow of the classic decomposition of generalization error into bias and variance, splitting uncertainty into its epistemic and aleatoric constituents separates an irreducible from a reducible part.
+
+The reducible part relates to imperfection in the model: In theory, if our model were perfect, epistemic uncertainty would vanish. Put differently, if the training data were unlimited -- or if they comprised the whole population -- we could just add capacity to the model until we've obtained a perfect fit.
+
+In contrast, normally there is variation in our measurements. There may be one true process that determines my resting heart rate; nonetheless, actual measurements will vary over time. There is nothing to be done about this: This is the aleatoric part that just remains, to be factored into our expectations.
+
+Now reading this, you might be thinking: "Wouldn't a model that actually _were_ perfect capture those pseudo-random fluctuations?". We'll leave that phisosophical question be;  instead, we'll try to illustrate the usefulness of this distinction by example, in a practical way. In a nutshell, viewing a model's _aleatoric_ uncertainty output should caution us to factor in appropriate deviations when making our predictions, while inspecting _epistemic_ uncertainty should help us re-think the appropriateness of the chosen model.
+
+Now let's dive in and see how we may accomplish our goal with `tfprobability`. We start with the simulated dataset.
+
+## Uncertainty estimates on simulated data
+
+### Dataset
+
+
+We re-use the dataset from the Google TensorFlow Probability team's [blog post on the same subject]( https://medium.com/tensorflow/regression-with-probabilistic-layers-in-tensorflow-probability-e46ff5d37baf) ^[see also the corresponding [notebook](https://colab.research.google.com/github/tensorflow/probability/blob/master/tensorflow_probability/examples/jupyter_notebooks/Probabilistic_Layers_Regression.ipynb#scrollTo=5zCEYpzu7bDX)], with one exception: We extend the range of the independent variable a bit on the negative side, to better demonstrate the different methods' behaviors.
+
+Here is the data-generating process. We also get library loading out of the way. Like in the preceding posts on `tfprobability`, please use the development versions of `tensorflow` and `tfprobability` as well as `keras`. Call `install_tensorflow(version = "nightly")` to obtain a current nightly build of TensorFlow and TensorFlow Probability:
+
+
+```{r}
+# make sure we use the development versions of tensorflow, tfprobability and keras
+devtools::install_github("rstudio/tensorflow")
+devtools::install_github("rstudio/tfprobability")
+devtools::install_github("rstudio/keras")
+
+# and that we use a nightly build of TensorFlow and TensorFlow Probability
+tensorflow::install_tensorflow(version = "nightly")
+
+library(tensorflow)
+tf$compat$v1$enable_v2_behavior()
```

Comment:
Do you really need to `enable_v2_behavior()`? I guess `tfprobabilty` requires it? You might want to consider using TF 2.0 in subsequent posts, assuming the fixes that you need are being ported to TF 2.0, is that the case?

Reply:
In TFP, they just do this all the time since a few months... I think here too, it's the safest way.
Also, now that TF 2 beta is imminent, seems like it won't be long for TF 2 anyway...

Reply:
I wonder if you should consider adding a:

```
.onLoad <- function() {
   # if tf < 2.0
   tf$compat$v1$enable_v2_behavior()
}
```

To `tfprobability`, that way you can get rid of the `tf$compat$v1$enable_v2_behavior()` statement.

Reply:
Again I was unclear, `tfprobability` per se doesn't need it - the honest way to express this would probably be "it's the idiomatic way to use TF master nowadays". It 's also a way to make sure the code will (hopefully!) work with the final TF 2.

Reply:
I would then consider moving clean up the code a bit as follows:

```
library(tensorflow)
library(tfprobability)
library(keras)

library(dplyr)
library(tidyr)
library(ggplot2)

# make sure this code is compatible with TensorFlow 2.0
tf$compat$v1$enable_v2_behavior()
```

Reply:
done!